filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobFileSystemManager.scala

```diff
@@ -201,9 +201,52 @@ case class WSMBlobSasTokenGenerator(wsmClientProvider: WorkspaceManagerApiClient
     }
   }
 
- def getContainerResourceId(workspaceId: UUID, container: BlobContainerName, wsmAuth : String): Try[UUID] = {
-    val wsmResourceClient = wsmClientProvider.getResourceApi(wsmAuth)
-    wsmResourceClient.findContainerResourceId(workspaceId, container)
+  private val cachedContainerResourceIds = new mutable.HashMap[BlobContainerName, UUID]()
+
+ private def getContainerResourceId(workspaceId: UUID, container: BlobContainerName, wsmAuth : String): Try[UUID] = {
+   cachedContainerResourceIds.get(container) match {
+     case Some(id) => Try(id)
+     case _ => {
+       val wsmResourceClient = wsmClientProvider.getResourceApi(wsmAuth)
+       val resourceId = wsmResourceClient.findContainerResourceId(workspaceId, container)
+       resourceId.map(id => cachedContainerResourceIds.put(container, id)) //NB: Modifying cache state here.
+       cachedContainerResourceIds.get(container) match {
+         case Some(uuid) => Try(uuid)
+         case _ => Failure(new NoSuchElementException("Could not retrieve container resource ID from WSM"))
+       }
+     }
+   }
+  }
+
+  private def getWsmAuth: Try[String] = {
+    overrideWsmAuthToken match {
+      case Some(t) => Success(t)
+      case None => AzureCredentials.getAccessToken(None).toTry
+    }
+  }
+  private def parseTerraWorkspaceIdFromPath(blobPath: BlobPath): Try[UUID] = {
+    if (blobPath.container.value.startsWith("sc-")) Try(UUID.fromString(blobPath.container.value.substring(3)))
+    else Failure(new Exception("Could not parse workspace ID from storage container. Are you sure this is a file in a Terra Workspace?"))
+  }
+
+  /**
+   * If the provided blob path looks like it comes from a terra workspace, return an end point that, when called with GET
+   * and proper authentication, will return a sas token capable of accessing the container the blob path points to.
+   * @param blobPath A blob path of a file living in a blob container that WSM knows about (likely a workspace container).
+   *
+   * NOTE: This function makes two synchronous REST requests.
+   */
+  def getWSMSasFetchEndpoint(blobPath: BlobPath): Try[String] = {
+    val wsmEndpoint = wsmClientProvider.getBaseWorkspaceManagerUrl
+    val terraInfo: Try[WSMTerraCoordinates] = for {
+      workspaceId <- parseTerraWorkspaceIdFromPath(blobPath)
+      auth <- getWsmAuth
```

Comment:
Can we do this auth-getting inside `getContainerResourceId` so we're sure we don't run it unless necessary? Getting the token from the env takes a while.